Voluntary exit should fail gracefully when beacon node API endpoint is incorrect

When the beacon node API endpoint is incorrectly specified (note that I used the wrong port number), the voluntary exit command should fail with a helpful error message rather than a stack trace:

```
$ sudo -u teku ./build/install/teku/bin/teku voluntary-exit --epoch=24500 --validator-keys=/var/lib/teku/keys_medalla:/var/lib/teku/keys_medalla --beacon-node-api-endpoint=http://localhost:5555
Loading configuration...
java.lang.RuntimeException: Error communicating with Beacon Node API: Failed to connect to localhost/127.0.0.1:5555
        at tech.pegasys.teku.validator.remote.apiclient.OkHttpValidatorRestApiClient.executeCall(OkHttpValidatorRestApiClient.java:315)
        at tech.pegasys.teku.validator.remote.apiclient.OkHttpValidatorRestApiClient.get(OkHttpValidatorRestApiClient.java:262)
        at tech.pegasys.teku.validator.remote.apiclient.OkHttpValidatorRestApiClient.getFork(OkHttpValidatorRestApiClient.java:98)
        at tech.pegasys.teku.cli.subcommand.VoluntaryExitCommand.getFork(VoluntaryExitCommand.java:156)
        at tech.pegasys.teku.cli.subcommand.VoluntaryExitCommand.initialise(VoluntaryExitCommand.java:176)
        at tech.pegasys.teku.cli.subcommand.VoluntaryExitCommand.run(VoluntaryExitCommand.java:87)
        at picocli.CommandLine.executeUserObject(CommandLine.java:1919)
        at picocli.CommandLine.access$1100(CommandLine.java:145)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2332)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2326)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2291)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2159)
        at picocli.CommandLine.execute(CommandLine.java:2058)
        at tech.pegasys.teku.cli.BeaconNodeCommand.parse(BeaconNodeCommand.java:229)
        at tech.pegasys.teku.Teku.main(Teku.java:32)
Caused by: java.net.ConnectException: Failed to connect to localhost/127.0.0.1:5555
        at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.kt:297)
        at okhttp3.internal.connection.RealConnection.connect(RealConnection.kt:207)
        at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.kt:226)
        at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.kt:106)
        at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.kt:74)
        at okhttp3.internal.connection.RealCall.initExchange$okhttp(RealCall.kt:255)
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:32)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
        at okhttp3.internal.connection.RealCall.execute(RealCall.kt:154)
        at tech.pegasys.teku.validator.remote.apiclient.OkHttpValidatorRestApiClient.executeCall(OkHttpValidatorRestApiClient.java:311)
        ... 14 more
Caused by: java.net.ConnectException: Connection refused
        at java.base/sun.nio.ch.Net.pollConnect(Native Method)
        at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:579)
        at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:542)
        at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:597)
        at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:339)
        at java.base/java.net.Socket.connect(Socket.java:603)
        at okhttp3.internal.platform.Platform.connectSocket(Platform.kt:120)
        at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.kt:295)
        ... 30 more
```

The scenario when the remote port is open but is still incorrect should also be handled gracefully. This has affected at least one user on [the Teku Discord](https://discord.com/channels/697535391594446898/697539289042649190/778217594179026944):

```
$ sudo -u teku ./build/install/teku/bin/teku voluntary-exit --epoch=24500 --validator-keys=/var/lib/teku/keys_medalla:/var/lib/teku/keys_medalla --beacon-node-api-endpoint=http://localhost:9000
Loading configuration...
java.lang.RuntimeException: Error communicating with Beacon Node API: Unexpected status line: /multistream/1.0.0
        at tech.pegasys.teku.validator.remote.apiclient.OkHttpValidatorRestApiClient.executeCall(OkHttpValidatorRestApiClient.java:315)
        at tech.pegasys.teku.validator.remote.apiclient.OkHttpValidatorRestApiClient.get(OkHttpValidatorRestApiClient.java:262)
        at tech.pegasys.teku.validator.remote.apiclient.OkHttpValidatorRestApiClient.getFork(OkHttpValidatorRestApiClient.java:98)
        at tech.pegasys.teku.cli.subcommand.VoluntaryExitCommand.getFork(VoluntaryExitCommand.java:156)
        at tech.pegasys.teku.cli.subcommand.VoluntaryExitCommand.initialise(VoluntaryExitCommand.java:176)
        at tech.pegasys.teku.cli.subcommand.VoluntaryExitCommand.run(VoluntaryExitCommand.java:87)
        at picocli.CommandLine.executeUserObject(CommandLine.java:1919)
        at picocli.CommandLine.access$1100(CommandLine.java:145)
        at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2332)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2326)
        at picocli.CommandLine$RunLast.handle(CommandLine.java:2291)
        at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2159)
        at picocli.CommandLine.execute(CommandLine.java:2058)
        at tech.pegasys.teku.cli.BeaconNodeCommand.parse(BeaconNodeCommand.java:229)
        at tech.pegasys.teku.Teku.main(Teku.java:32)
Caused by: java.net.ProtocolException: Unexpected status line: /multistream/1.0.0
        at okhttp3.internal.http.StatusLine$Companion.parse(StatusLine.kt:80)
        at okhttp3.internal.http1.Http1ExchangeCodec.readResponseHeaders(Http1ExchangeCodec.kt:178)
        at okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.kt:106)
        at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.kt:79)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:34)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
        at okhttp3.internal.connection.RealCall.execute(RealCall.kt:154)
        at tech.pegasys.teku.validator.remote.apiclient.OkHttpValidatorRestApiClient.executeCall(OkHttpValidatorRestApiClient.java:311)
        ... 14 more
```

--- a/teku/src/main/java/tech/pegasys/teku/cli/subcommand/VoluntaryExitCommand.java
+++ b/teku/src/main/java/tech/pegasys/teku/cli/subcommand/VoluntaryExitCommand.java
@@ -14,6 +14,7 @@
 package tech.pegasys.teku.cli.subcommand;
 
 import com.google.common.base.Throwables;
+import java.io.UncheckedIOException;
 import java.net.ConnectException;
 import java.net.URI;
 import java.nio.charset.Charset;
@@ -88,9 +89,16 @@ public class VoluntaryExitCommand implements Runnable {
   @Override
   public void run() {
     SUB_COMMAND_LOG.display("Loading configuration...");
-    initialise();
-    confirmExits();
-    getValidatorIndices(blsPublicKeyValidatorMap).forEach(this::submitExitForValidator);
+    try {
+      initialise();
+      confirmExits();
+      getValidatorIndices(blsPublicKeyValidatorMap).forEach(this::submitExitForValidator);
+    } catch (UncheckedIOException ex) {
+      if (Throwables.getRootCause(ex) instanceof ConnectException) {
+        SUB_COMMAND_LOG.error(getFailedToConnectMessage());
+        System.exit(1);
+      }
+    }
   }
 
   private void confirmExits() {
@@ -157,27 +165,11 @@ public class VoluntaryExitCommand implements Runnable {
   }
 
   private Optional<Bytes32> getGenesisRoot() {
-    try {
-      return apiClient.getGenesis().map(response -> response.getData().getGenesisValidatorsRoot());
-    } catch (RuntimeException ex) {
-      if (Throwables.getRootCause(ex) instanceof ConnectException) {
-        SUB_COMMAND_LOG.error(getFailedToConnectMessage());
-        System.exit(1);
-      }
-    }
-    return Optional.empty();
+    return apiClient.getGenesis().map(response -> response.getData().getGenesisValidatorsRoot());
   }
 
   private Optional<tech.pegasys.teku.datastructures.state.Fork> getFork() {
-    try {
-      return apiClient.getFork().map(Fork::asInternalFork);
-    } catch (RuntimeException ex) {
-      if (Throwables.getRootCause(ex) instanceof ConnectException) {
-        SUB_COMMAND_LOG.error(getFailedToConnectMessage());
-        System.exit(1);
-      }
-    }
-    return Optional.empty();
+    return apiClient.getFork().map(Fork::asInternalFork);
   }
 
   private void initialise() {

--- a/teku/src/main/java/tech/pegasys/teku/cli/subcommand/VoluntaryExitCommand.java
+++ b/teku/src/main/java/tech/pegasys/teku/cli/subcommand/VoluntaryExitCommand.java
@@ -13,6 +13,9 @@
 
 package tech.pegasys.teku.cli.subcommand;
 
+import com.google.common.base.Throwables;
+import java.net.ConnectException;
+import java.net.URI;
 import java.nio.charset.Charset;
 import java.nio.file.Path;
 import java.util.HashMap;
@@ -66,6 +69,7 @@ public class VoluntaryExitCommand implements Runnable {
   private tech.pegasys.teku.datastructures.state.Fork fork;
   private Bytes32 genesisRoot;
   private Map<BLSPublicKey, Validator> blsPublicKeyValidatorMap;
+  private TekuConfiguration config;
 
   @CommandLine.Mixin(name = "Validator Keys")
   private ValidatorKeysOptions validatorKeysOptions;
@@ -149,15 +153,31 @@ public class VoluntaryExitCommand implements Runnable {
   }
 
   private Optional<Bytes32> getGenesisRoot() {
-    return apiClient.getGenesis().map(response -> response.getData().getGenesisValidatorsRoot());
+    try {
+      return apiClient.getGenesis().map(response -> response.getData().getGenesisValidatorsRoot());
+    } catch (RuntimeException ex) {
+      if (Throwables.getRootCause(ex) instanceof ConnectException) {
+        SUB_COMMAND_LOG.error(getFailedToConnectMessage());
+        System.exit(1);
+      }
+    }
+    return Optional.empty();
   }
 
   private Optional<tech.pegasys.teku.datastructures.state.Fork> getFork() {
-    return apiClient.getFork().map(Fork::asInternalFork);
+    try {
+      return apiClient.getFork().map(Fork::asInternalFork);
+    } catch (RuntimeException ex) {
+      if (Throwables.getRootCause(ex) instanceof ConnectException) {
+        SUB_COMMAND_LOG.error(getFailedToConnectMessage());
+        System.exit(1);
+      }
+    }
+    return Optional.empty();
   }
 
   private void initialise() {
-    final TekuConfiguration config = tekuConfiguration();
+    config = tekuConfiguration();
     final AsyncRunnerFactory asyncRunnerFactory =
         new AsyncRunnerFactory(new MetricTrackingExecutorFactory(new NoOpMetricsSystem()));
     final AsyncRunner asyncRunner = asyncRunnerFactory.create("voluntary-exits", 8);
@@ -211,6 +231,17 @@ public class VoluntaryExitCommand implements Runnable {
     builder.data(config -> config.dataBasePath(Path.of(".")));
     builder.validator(config -> config.validatorKeystoreLockingEnabled(false));
     return builder.build();
+  }
+
+  private String getFailedToConnectMessage() {
+    return String.format(
+        "Failed to connect to beacon node. Check that %s is available.",
+        config
+            .validatorClient()
+            .getValidatorConfig()
+            .getBeaconNodeApiEndpoint()
+            .orElse(URI.create("http://127.0.0.1:5051"))
+            .toString());
   }
 
   private void buildGlobalConfiguration(final GlobalConfigurationBuilder builder) {

--- a/validator/remote/src/main/java/tech/pegasys/teku/validator/remote/apiclient/OkHttpValidatorRestApiClient.java
+++ b/validator/remote/src/main/java/tech/pegasys/teku/validator/remote/apiclient/OkHttpValidatorRestApiClient.java
@@ -33,6 +33,7 @@ import static tech.pegasys.teku.validator.remote.apiclient.ValidatorApiMethod.SU
 
 import com.fasterxml.jackson.core.JsonProcessingException;
 import java.io.IOException;
+import java.io.UncheckedIOException;
 import java.util.Collection;
 import java.util.Collections;
 import java.util.HashMap;
@@ -312,7 +313,8 @@ public class OkHttpValidatorRestApiClient implements ValidatorRestApiClient {
       LOG.trace("{} {} {}", request.method(), request.url(), response.code());
       return responseHandler.handleResponse(request, response);
     } catch (IOException e) {
-      throw new RuntimeException("Error communicating with Beacon Node API: " + e.getMessage(), e);
+      throw new UncheckedIOException(
+          "Error communicating with Beacon Node API: " + e.getMessage(), e);
     }
   }
 

